### PR TITLE
Auth back end

### PR DESCRIPTION
**Pull request by AlexRSc**
Add Registration functionality, AuthService, UserRepo, UserEntity and one IntegrationTest so far. 

Next ErrorHandling, JWT+SecurityConfig, 2 more tests for Registration and Login BackEnd Functionality

### CODE REVIEW

File: backend/pom.xml [hunk 1]
```diff
@@ -38,7 +38,70 @@
...
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-configuration-processor</artifactId>
+        </dependency>
+        <!-- Swagger -->
+        <dependency>
+            <groupId>io.springfox</groupId>
+            <artifactId>springfox-boot-starter</artifactId>
+            <version>3.0.0</version>
+        </dependency>
```
Comment (jamarob) [inline backend/pom.xml:61]: PR description says nothing about swagger, not a big thing, but try to keep your prs focuseds.

File: backend/src/main/java/de/neuefische/CapStone/backend/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import static org.springframework.http.ResponseEntity.ok;
+
+@Api(
+        tags = AuthController.AUTH_CONTROLLER_TAG
+)
+@RestController
+@CrossOrigin
+@RequestMapping("/user")
+@Getter
+@Setter
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/CapStone/backend/controller/AuthController.java:27]:
```suggestion
```
`AuthService` is final, no need for setters and getters

File: backend/src/main/java/de/neuefische/CapStone/backend/model/UserEntity.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package de.neuefische.CapStone.backend.model;
+
+import lombok.*;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name="smart_user")
+@Getter
+@Setter
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/CapStone/backend/model/UserEntity.java]:
```suggestion
@Data
```

File: backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Optional;
+
+import static org.springframework.util.StringUtils.hasText;
+
+@Service
+@Getter
+@Setter
+public class AuthService {
+
+    private UserRepository userRepository;
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java]:
```suggestion
public class AuthService {

    private final UserRepository userRepository;
```

File: backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            throw new IllegalArgumentException("Name must not be blank");
+        }
+        checkIfUserExists(name);
+        String email = userEntity.getEmail();
+        if(!hasText(email)) {
+            throw new IllegalArgumentException("Email must not be blank");
+        }
+        checkIfEmailExists(email);
+        userRepository.save(userEntity);
+        return userEntity;
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java]:
```suggestion
        return userRepository.save(userEntity);
```
Always return the saved entity

File: backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    }
+
+    private void checkIfEmailExists(String email) {
+        Optional<UserEntity> existingEmail = userRepository.findByEmail(email);
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java]: Maybe you could use a `existsByEmail` query method to the repository?

File: backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        if(existingEmail.isPresent()){
+            throw new EntityExistsException(String.format("Email with name %s already exists!", email));
+        }
+    }
+
+    private void checkIfUserExists(String name) {
+        Optional<UserEntity> existingUser = userRepository.findByUserName(name);
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java]: Same as above with `existsByUserName`

File: backend/src/test/java/de/neuefische/CapStone/backend/controller/AuthControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                .password("ImInLoveWithMario").build();
+        String newUrl = url() + "/user/registration";
+        //WHEN
+        HttpEntity<User> httpEntity = new HttpEntity<>(userToAdd);
+        ResponseEntity<User> response = testRestTemplate.exchange(newUrl, HttpMethod.POST, httpEntity, User.class);
+
+        //THEN
+        assertThat(response.getStatusCode(), is(HttpStatus.OK));
+        assertThat(response.getBody().getUserName(), is(userToAdd.getUserName()));
+        assertThat(response.getBody().getEmail(), is(userToAdd.getEmail()));
```
Comment (jamarob) [inline backend/src/test/java/de/neuefische/CapStone/backend/controller/AuthControllerTest.java]: Maybe assert that the user was really added to the database?

**Review by jamarob — APPROVED** [review]
Good job and easy to review 👍 

I am wondering about the docker stuff in this pr..?

File: backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java (hunk 6 above)
Comment (AlexRSc) [reply backend/src/main/java/de/neuefische/CapStone/backend/service/AuthService.java]: > Good job and easy to review 👍
> 
> I am wondering about the docker stuff in this pr..?

Thanks for the review, I implemented all of those thing. The Docker Stuff was part of my "Oh no, I don´t know why my database isn´t working"-reaction, but its necessary (I think). I should have added that into the PR-comment though.

**Pull request merged**